File: src/com/peyo/launcherlb/LauncherModel.java
package com.peyo.launcherlb;

import java.util.ArrayList;
import java.util.List;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.os.Handler;
import android.os.HandlerThread;

public class LauncherModel {
    public interface Callbacks {
        void bindAllApplications(ArrayList<AppInfo> apps);
    }
    
    static final HandlerThread sWorkerThread = new HandlerThread("launcher-loader");
    static {
        sWorkerThread.start();
    }
    static final Handler sWorker = new Handler(sWorkerThread.getLooper());

    private final Context mContext;
    private Callbacks mCallbacks;
    private ArrayList<AppInfo> mAppList;
    DeferredHandler mHandler = new DeferredHandler();

    public LauncherModel(Context context, Callbacks callbacks) {
        mContext  = context;
        mAppList = new ArrayList<>();
        mCallbacks = callbacks;
    }

    public void startLoader() {
        sWorker.post(new Runnable() {
            @Override
            public void run() {
                getAppList();
                mHandler.post(new Runnable() {
                    public void run() {
                        if (mCallbacks != null) {
                            mCallbacks.bindAllApplications(mAppList);
                        }
                    }
                });
            }
        });
    }

    private static final String CATEGORY_LEANBACK_SETTINGS =
            "android.intent.category.LEANBACK_SETTINGS";

    private void getAppList() {
        mAppList.clear();
        PackageManager pm = mContext.getPackageManager();
        List<ResolveInfo> rInfos = pm.queryIntentActivities(
                new Intent(Intent.ACTION_MAIN).addCategory(Intent.CATEGORY_LAUNCHER), 0);
        for (ResolveInfo ri : rInfos) {
            AppInfo ai = new AppInfo(mContext, ri.activityInfo);
            mAppList.add(ai);
        }

        rInfos = pm.queryIntentActivities(
                new Intent(Intent.ACTION_MAIN).addCategory(Intent.CATEGORY_LEANBACK_LAUNCHER),0);
        for (ResolveInfo ri : rInfos) {
            AppInfo ai = new AppInfo(mContext, ri.activityInfo);
            addToAppList(ai);
        }

        rInfos = pm.queryIntentActivities(
                new Intent(Intent.ACTION_MAIN).addCategory(CATEGORY_LEANBACK_SETTINGS),0);
        for (ResolveInfo ri : rInfos) {
            AppInfo ai = new AppInfo(mContext, ri.activityInfo);
            if (!ai.getComponentName().getClassName().contains("NetworkActivity")) {
                mAppList.add(ai);
            }
        }
    }

    private void addToAppList(AppInfo info) {
        for (AppInfo ai: mAppList) {
            if (info.getComponentName().equals(ai.getComponentName())) {
                return;
            }
        }
        mAppList.add(info);
    }
}

File: src/com/peyo/rplauncher/LauncherModel.kt
package com.peyo.rplauncher

import android.content.Context
import android.content.Intent
import android.os.Handler
import android.os.HandlerThread
import java.util.ArrayList

class LauncherModel(private val mContext: Context, private val mCallbacks: Callbacks?) {
    private val mAppList: ArrayList<AppInfo>
    internal var mHandler = DeferredHandler()

    interface Callbacks {
        fun bindAllApplications(apps: ArrayList<AppInfo>)
    }

    init {
        mAppList = ArrayList()
    }

    fun startLoader() {
        sWorker.post {
            getAppList()
            mHandler.post(Runnable {
                mCallbacks?.bindAllApplications(mAppList)
            })
        }
    }

    private fun getAppList() {
        mAppList.clear()
        val pm = mContext.packageManager
        var rInfos = pm.queryIntentActivities(
                Intent(Intent.ACTION_MAIN).addCategory(Intent.CATEGORY_LAUNCHER), 0)
        for (ri in rInfos) {
            val ai = AppInfo(mContext, ri.activityInfo)
            mAppList.add(ai)
        }

        rInfos = pm.queryIntentActivities(
                Intent(Intent.ACTION_MAIN).addCategory(Intent.CATEGORY_LEANBACK_LAUNCHER), 0)
        for (ri in rInfos) {
            val ai = AppInfo(mContext, ri.activityInfo)
            addToAppList(ai)
        }

        rInfos = pm.queryIntentActivities(
                Intent(Intent.ACTION_MAIN).addCategory(CATEGORY_LEANBACK_SETTINGS), 0)
        for (ri in rInfos) {
            val ai = AppInfo(mContext, ri.activityInfo)
            if (!ai.componentName.className.contains("NetworkActivity")) {
                mAppList.add(ai)
            }
        }
    }

    private fun addToAppList(info: AppInfo) {
        for (ai in mAppList) {
            if (info.componentName == ai.componentName) {
                return
            }
        }
        mAppList.add(info)
    }

    companion object {

        internal val sWorkerThread = HandlerThread("launcher-loader")

        init {
            sWorkerThread.start()
        }

        internal val sWorker = Handler(sWorkerThread.looper)

        private val CATEGORY_LEANBACK_SETTINGS = "android.intent.category.LEANBACK_SETTINGS"
    }
}